### PR TITLE
Use eslint with loopback config

### PR DESCRIPTION
**Pull request by bajtos**
Add eslint, fix style violations.

See also strongloop/eslint-config-loopback#1 and strongloop/eslint-config-loopback#2

Connect to strongloop/loopback#1749

@superkhau please review
/cc @strongloop/loopback-dev

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+  "extends": "loopback",
+  "rules": {
+    "max-len": ["error", 110, 4, {
+      "ignoreComments": true,
+      "ignoreUrls": true,
+      "ignorePattern": "^\\s*var\\s.+=\\s*(require\\s*\\()|(/)"
+    }],
+    // NOTE(bajtos) we should eventuall remove this override
+    // and fix all of those 100+ violations
+    "one-var": "off"
```
Comment (bajtos) [inline .eslintrc]: This is a bit of a shortcut I took in order to get this landed reasonably soon. There were already too many changes needed to get eslint to pass (well over 400), these extra 100+ were simply too much to do in one go.

File: test/.eslintrc [hunk 2]
```diff
@@ -0,0 +1,7 @@
+{
+  "rules": {
+    // NOTE(bajtos) We should eventually get rid of this override,
+    // and fix those 200+ violations instead
+    "padded-blocks": "off"
```
Comment (bajtos) [inline test/.eslintrc]: And ditto here.

File: .eslintrc (hunk 1 above)
Comment (superkhau) [inline .eslintrc]: Good idea. I assume you're using --fix?

File: lib/model.js [hunk 3]
```diff
@@ -318,15 +321,15 @@ ModelBaseClass.prototype._initProperties = function (data, options) {
...
+        if (!(self.__data[p] instanceof type) &&
+            typeof self.__data[p] === 'object' &&
+            self.__data[p] !== null) {
           self.__data[p] = new type(self.__data[p]);
         }
       } else if (type.name === 'Array' || Array.isArray(type)) {
-        if (!(self.__data[p] instanceof List)
-          && self.__data[p] !== undefined
-          && self.__data[p] !== null ) {
+        if (!(self.__data[p] instanceof List) &
```
Comment (superkhau) [inline lib/model.js]: &&?

File: lib/relation-definition.js [hunk 4]
```diff
@@ -787,9 +787,9 @@ HasMany.prototype.findById = function (fkId, options, cb) {
     if (inst[fk] != null && idEquals(inst[fk], modelInstance[pk])) {
       cb(null, inst);
     } else {
-      err = new Error('Key mismatch: ' + modelFrom.modelName + '.' + pk
-        + ': ' + modelInstance[pk]
-        + ', ' + modelTo.modelName + '.' + fk + ': ' + inst[fk]);
+      err = new Error('Key mismatch: ' + modelFrom.modelName + '.' + pk +
+        ': ' + modelInstance[pk] +
+        ', ' + modelTo.modelName + '.' + fk + ': ' + inst[fk]);
```
Comment (superkhau) [inline lib/relation-definition.js]: join with line above? lots of empty space after

File: lib/relation-definition.js [hunk 5]
```diff
@@ -2412,11 +2411,11 @@ EmbedsMany.prototype.prepareEmbeddedInstance = function(inst) {
...
-  modelInstance = modelInstance || this.modelInstance;
-  var embeddedList = modelInstance[this.definition.keyFrom] || [];
-  embeddedList.forEach(this.prepareEmbeddedInstance.bind(this));
-  return embeddedList;
-};
+    modelInstance = modelInstance || this.modelInstance;
+    var embeddedList = modelInstance[this.definition.keyFrom] || [];
+    embeddedList.forEach(this.prepareEmbeddedInstance.bind(this));
+    return embeddedList;
+  };
```
Comment (superkhau) [inline lib/relation-definition.js]: remove indents in front from 2413 to 2418 to match how we define other funcs. ie)

```
Embeds... =
Embeds... = func() {
  ...
};
```

File: lib/relation-definition.js [hunk 6]
```diff
@@ -2540,8 +2539,8 @@ EmbedsMany.prototype.updateById = function (fkId, data, options, cb) {
     if (typeof cb === 'function') {
       modelInstance.updateAttribute(propertyName,
         embeddedList, options, function(err) {
```
Comment (superkhau) [inline lib/relation-definition.js]: move this stuff up a line and fix indents after

File: lib/relation-definition.js [hunk 7]
```diff
@@ -2571,9 +2570,9 @@ EmbedsMany.prototype.destroyById = function (fkId, options, cb) {
     if (typeof cb === 'function') {
       modelInstance.updateAttribute(propertyName,
         embeddedList, function(err) {
```
Comment (superkhau) [inline lib/relation-definition.js]: ditto

File: lib/relation-definition.js [hunk 8]
```diff
@@ -2608,8 +2607,8 @@ EmbedsMany.prototype.destroyAll = function(where, options, cb) {
   if (typeof cb === 'function') {
     modelInstance.updateAttribute(propertyName,
       embeddedList, function(err) {
```
Comment (superkhau) [inline lib/relation-definition.js]: ditto

File: lib/relation-definition.js [hunk 9]
```diff
@@ -2664,13 +2663,13 @@ EmbedsMany.prototype.create = function (targetModelData, options, cb) {
     if (modelInstance.isNewRecord()) {
       modelInstance.setAttribute(propertyName, embeddedList);
       modelInstance.save(options, function(err) {
-          cb(err, err ? null : inst);
+        cb(err, err ? null : inst);
       });
     } else {
       modelInstance.updateAttribute(propertyName,
         embeddedList, options, function(err) {
```
Comment (superkhau) [inline lib/relation-definition.js]: ditto

File: lib/relation-definition.js [hunk 10]
```diff
@@ -3029,11 +3028,11 @@ ReferencesMany.prototype.exists = function (fkId, options, cb) {
   var ids = this.modelInstance[fk] || [];
 
   cb = cb || utils.createPromiseCallback();
-  process.nextTick(function() { cb(null, utils.findIndexOf(ids, fkId, idEquals) > -1) });
+  process.nextTick(function() { cb(null, utils.findIndexOf(ids, fkId, idEquals) > -1); });
```
Comment (superkhau) [inline lib/relation-definition.js]:
```
process.nextTick(function() {
  cb(null, utils.findIndexOf(ids, fkId, idEquals) > -1);
});
```

File: lib/relation-definition.js [hunk 11]
```diff
@@ -3125,8 +3124,8 @@ ReferencesMany.prototype.create = function (targetModelData, options, cb) {
 
     modelInstance.updateAttribute(fk,
       ids, options, function(err, modelInst) {
```
Comment (superkhau) [inline lib/relation-definition.js]: ditto

File: lib/scope.js [hunk 12]
```diff
@@ -277,47 +277,47 @@ function defineScope(cls, targetClass, name, params, methods, options) {
 
   cls['__get__' + name] = fn;
 
-  var fn_create = function () {
+  var fnCreate = function() {
     var f = this[name].create;
     f.apply(this[name], arguments);
   };
```
Comment (superkhau) [inline lib/scope.js]:
```
function createFunction() {
   var f = this[name].create;
   f.apply(this[name], arguments);
}
```

Then just assign create:

`cls['__create__' + name] = createFunction;`

Comment (superkhau) [inline lib/scope.js]: Can do for the rest, but its a nitpick anyways

File: lib/types.js [hunk 13]
```diff
@@ -9,7 +9,7 @@ Types.Text = function Text(value) {
   this.value = value;
 }; // Text type
 
-Types.Text.prototype.toObject = Types.Text.prototype.toJSON = function () {
+Types.Text.prototype.toObject = Types.Text.prototype.toJSON = function() {
```
Comment (superkhau) [inline lib/types.js]: These should be broken out into two lines:

```
Types.Text.prototype.toObject =
Types.Text.prototype.toJSON = function() {
```

File: lib/types.js [hunk 14]
```diff
@@ -19,7 +19,7 @@ Types.JSON = function JSON(value) {
   }
   this.value = value;
 }; // JSON Object
-Types.JSON.prototype.toObject = Types.JSON.prototype.toJSON = function () {
+Types.JSON.prototype.toObject = Types.JSON.prototype.toJSON = function() {
```
Comment (superkhau) [inline lib/types.js]: this too

File: lib/types.js [hunk 15]
```diff
@@ -29,20 +29,19 @@ Types.Any = function Any(value) {
   }
   this.value = value;
 }; // Any Type
-Types.Any.prototype.toObject = Types.Any.prototype.toJSON = function () {
+Types.Any.prototype.toObject = Types.Any.prototype.toJSON = function() {
```
Comment (superkhau) [inline lib/types.js]: this too

File: support/describe-operation-hooks.js [hunk 16]
```diff
@@ -12,12 +12,12 @@ var Memory = require('../lib/connectors/memory').Memory;
 var HOOK_NAMES = [
   'access',
   'before save', 'persist', 'loaded', 'after save',
-  'before delete', 'after delete'
+  'before delete', 'after delete',
```
Comment (superkhau) [inline support/describe-operation-hooks.js]: should be a single line per item vertical list in alphabetical order

File: test/CustomTypeForeignKey.test.js [hunk 17]
```diff
@@ -4,36 +4,36 @@ var jdb = require('../');
...
-          "dataType": "string"
-        }
-      }
+      'myProp': {
+        'id': true,
+        'type': 'string',
+        'memory': {
+          'dataType': 'string',
+        },
+      },
```
Comment (superkhau) [inline test/CustomTypeForeignKey.test.js]: single quotes not need on key values

File: test/CustomTypeForeignKey.test.js [hunk 18]
```diff
@@ -4,36 +4,36 @@ var jdb = require('../');
...
-          "foreignKey": "myProp"
-        }
-      }
+        'item': {
+          'type': 'belongsTo',
+          'as': 'item',
+          'model': 'Item',
+          'foreignKey': 'myProp',
+        },
+      },
```
Comment (superkhau) [inline test/CustomTypeForeignKey.test.js]: same here

File: test/basic-querying.test.js [hunk 19]
```diff
@@ -215,53 +215,53 @@ describe('basic-querying', function () {
...
         done();
       });
     });
 
-    it('should support date "gte" that is satisfied', function (done) {
-      User.find({order: 'seq', where: { birthday: { "gte": new Date('1980-12-08') }
-      }}, function (err, users) {
+    it('should support date "gte" that is satisfied', function(done) {
+      User.find({ order: 'seq', where: { birthday: { 'gte': new Date('1980-12-08') },
+      }}, function(err, users) {
```
Comment (superkhau) [inline test/basic-querying.test.js]: }} should be on line above, this line should start with function

File: test/basic-querying.test.js [hunk 20]
```diff
@@ -432,9 +432,9 @@ describe('basic-querying', function () {
       });
     });
 
-    it('should support boolean "lt" that is satisfied', function (done) {
-      User.find({order: 'seq', where: { vip: { "lt": true }
-      }}, function (err, users) {
+    it('should support boolean "lt" that is satisfied', function(done) {
+      User.find({ order: 'seq', where: { vip: { 'lt': true },
+      }}, function(err, users) {
```
Comment (superkhau) [inline test/basic-querying.test.js]: all the functions up to here have the same issue

File: test/basic-querying.test.js [hunk 21]
```diff
@@ -462,11 +462,11 @@ describe('basic-querying', function () {
                 done();
               }
 
-              users.forEach(function (user) {
+              users.forEach(function(user) {
                 var obj = user.toObject();
 
                 Object.keys(obj)
-                  .forEach(function (key) {
+                  .forEach(function(key) {
```
Comment (superkhau) [inline test/basic-querying.test.js]: this forEach can be moved up a line and then we can remove on level of nesting

**Comment by superkhau** [comment]
LGTM for as much as GH would show me.

File: .eslintrc (hunk 1 above)
Comment (bajtos) [inline .eslintrc]: Yes, I am using `--fix`, unfortunately not all rules implement auto-fixing :(

File: lib/model.js (hunk 3 above)
Comment (bajtos) [inline lib/model.js]: good catch!

File: lib/relation-definition.js (hunk 4 above)
Comment (bajtos) [inline lib/relation-definition.js]: I am trying to minimise the amount of changes, let's not apply your proposal.

File: lib/relation-definition.js (hunk 5 above)
Comment (bajtos) [inline lib/relation-definition.js]: good catch, will do.

File: lib/relation-definition.js (hunk 6 above)
Comment (bajtos) [inline lib/relation-definition.js]: I am rather reluctant to change things that don't need to be changed, because unnecessary whitespace/formatting changes make code archaeology difficult. Additionally, I cannot fit all arguments on a single 80 char line :(

File: lib/scope.js (hunk 12 above)
Comment (bajtos) [inline lib/scope.js]: I think this is to prevent collision with `DataAccessObject.create = function create(...) { ... }`.

File: lib/types.js (hunk 13 above)
Comment (bajtos) [inline lib/types.js]: Let's not make changes that are not absolutely necessary.

File: support/describe-operation-hooks.js (hunk 16 above)
Comment (bajtos) [inline support/describe-operation-hooks.js]: In this place, I am grouping logically-related hooks together. Also note that the order here dictates the order of columns in the output table, where I definitely want to see hooks listed in an order that makes it easy to compare context properties reported by related hooks.

File: test/CustomTypeForeignKey.test.js (hunk 17 above)
Comment (bajtos) [inline test/CustomTypeForeignKey.test.js]: will fix

File: test/basic-querying.test.js (hunk 20 above)
Comment (bajtos) [inline test/basic-querying.test.js]: Yeah, I know this style is not what we would like to see, but please let's not get too strict about existing legacy code. As I mentioned before, I'd like to keep the amount of changes as small as possible, to make it easier to use  `git blame`.

File: lib/relation-definition.js (hunk 6 above)
Comment (superkhau) [inline lib/relation-definition.js]: I don't mind leaving whitespace/formatting/indenting till we put everything through the linter. It'll be much easier to fix those when all the other rules clean up the majority of the code. Technically, if you can't fit it on a line I assume the code should look like:

```
modelInstance.updateAttribute(propertyName, embeddedList, options,
function(err) {
  ...
});
```

However, the ROI for fixing indents/whitespace/formatting is not key for this pass. I do believe we should fix these as we see them and as we work on related features.

File: lib/scope.js (hunk 12 above)
Comment (superkhau) [inline lib/scope.js]: Makes sense. But then we could still do:

```
function createFn() {
}

cls['__create__' + name] = createFn;
```
ROI is low here anyways, I hate to say it but I guess we follow "if it ain't broke, don't fix it" here. ;)
